[Opt Overlap] Create Optimizer Hook State directly from functional optim

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71620
* #71608
* #71607
* #71606
* #71605
* #71604
* #71603
* __->__ #71602

The design in https://github.com/pytorch/pytorch/issues/67570 requires
`_OptimizerHookState` to be created directly from a functional optimizer. Add
support and tests for this. Now, `OptimizerHookState` can be created with an already instantiated `_FunctionalOptimizer` with `OptimizerHookState.from_functional_optim`. Also refactor a few tests.

Differential Revision: [D33687477](https://our.internmc.facebook.com/intern/diff/D33687477/)

rohan-varma committed:
[Opt Overlap] Create Optimizer Hook State directly from functional optim

The design in https://github.com/pytorch/pytorch/issues/67570 requires
`_OptimizerHookState` to be created directly from a functional optimizer. Add
support and tests for this. Also refactor a few tests.

Differential Revision: [D33687477](https://our.internmc.facebook.com/intern/diff/D33687477/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:
```diff
@@ -21,9 +22,20 @@ def __init__(
+        self._check_valid_functional_optim()
+
+    @classmethod
+    def from_functional_optim(cls, functional_optim):
```
The difference between this function and the regular `__init__` is that this directly takes a functional optim class, but adding more arguments to the init and conditionally picking which one to use seemed a bit hacky.

From some online resources, it seems that one pythonic way around this is to use classmethods as here, hence this approach.

cbalioglu replied:
A nit question that we don't have to address in this PR. Would it be a BC-breaking change if we just supported constructing `OptimizerHookState` directly from an optim class? What do we gain by the current `__init__` behavior beyond "injecting" our `_allow_empty_param_list` argument?

rohan-varma committed:
Update on "[Opt Overlap] Create Optimizer Hook State directly from functional optim"


The design in https://github.com/pytorch/pytorch/issues/67570 requires
`_OptimizerHookState` to be created directly from a functional optimizer. Add
support and tests for this. Now, `OptimizerHookState` can be created with an already instantiated `_FunctionalOptimizer` with `OptimizerHookState.from_functional_optim`. Also refactor a few tests.

Differential Revision: [D33687477](https://our.internmc.facebook.com/intern/diff/D33687477/)

[ghstack-poisoned]

cbalioglu commented on torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:
```diff
@@ -21,9 +22,20 @@ def __init__(
+    @classmethod
+    def from_functional_optim(cls, functional_optim):
+        opt_hook_state_inst = cls.__new__(cls)  # Does not call __init__
+        opt_hook_state_inst.functional_optimizer = functional_optim
```
(update: saw your unit test below that sets `_allow_empty_param_list` externally, please see my comment for that as well)

Should we set `_allow_empty_param_list` somewhere similar to `__init__`?

rohan-varma replied:
Responded on the unittest comment!

cbalioglu commented on torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:
```diff
@@ -21,9 +22,20 @@ def __init__(
+        return opt_hook_state_inst
+
+    def _check_valid_functional_optim(self):
+        if not hasattr(self.functional_optimizer, _FUNCTIONAL_OPTIM_STEP_METHOD_NAME):
```
Should we add a brief docstr to `from_functional_optim` that describes this requirement? I see that we already mention this in `__init__`'s description.

rohan-varma replied:
Added this in the latest update, although I later realized that I should just enhance the doc in the top diff, as this is just moved to a regular `__init__` at the top diff.

cbalioglu commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -60,11 +65,13 @@
     DistTestCases
 )
 from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
```
Typo? `parameterized`? :)

rohan-varma replied:
It looks like both are used, but parameterized is more common: https://wikidiff.com/parameterize/parametrize

Although, this is already an implemented helper function that I'm just importing and using, so should be okay to leave it as is

cbalioglu commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -60,11 +65,13 @@
     FILE_SCHEMA,
     IS_FBCODE,
     NO_MULTIPROCESSING_SPAWN,
+    parametrize,
```
Or maybe not? I am not a linguistic expert so not sure whether it should be called parametrize or parameterize :)

cbalioglu commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3944,11 +3952,22 @@ def _test_ddp_hook_with_optimizer_parity(
+                            [],
+                            *functional_optim_args,
+                            **functional_optim_kwargs,
+                            _allow_empty_param_list=True
```
Oh, I see that you set `_allow_empty_param_list` before calling the factory function. In my opinion this is an internal implementation detail that the user should not be aware of. Is it possible to enforce this condition later inside `from_functional_optim`?

rohan-varma replied:
Yes, this is constructing the functional optimizer. Eventually this construction logic is moved into `as_functional_optim`, and called by i.e. L60 in this PR: https://github.com/pytorch/pytorch/pull/71605/files. 

I think by the end of this stack, `as_functional_optim` which sets `_allow_empty_param_list` is only called in construction of `_OverlappedOptimizer` which is called by DDP, so the end user is not expected to interface with it.

It would be a little tricky to move it to `OptimizerHookState.from_functional_optim`, because `_allow_empty_param_list` needs to be specified in the constructor. We could refactor `OptimizerHookState.from_functional_optim` to construct the entire functional optimizer, but I don't think either approach will make too much of a difference as neither API is intended to be public and both would be called from the same location in `_OverlappedOptimizer` anyways.

cbalioglu reviewed (commented):
Just a few minor questions before approving. Otherwise looks good to me.

rohan-varma committed:
Update on "[Opt Overlap] Create Optimizer Hook State directly from functional optim"


The design in https://github.com/pytorch/pytorch/issues/67570 requires
`_OptimizerHookState` to be created directly from a functional optimizer. Add
support and tests for this. Now, `OptimizerHookState` can be created with an already instantiated `_FunctionalOptimizer` with `OptimizerHookState.from_functional_optim`. Also refactor a few tests.

Differential Revision: [D33687477](https://our.internmc.facebook.com/intern/diff/D33687477/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Opt Overlap] Create Optimizer Hook State directly from functional optim"


The design in https://github.com/pytorch/pytorch/issues/67570 requires
`_OptimizerHookState` to be created directly from a functional optimizer. Add
support and tests for this. Now, `OptimizerHookState` can be created with an already instantiated `_FunctionalOptimizer` with `OptimizerHookState.from_functional_optim`. Also refactor a few tests.

Differential Revision: [D33687477](https://our.internmc.facebook.com/intern/diff/D33687477/)

[ghstack-poisoned]
